egs/librispeech/ASR/conformer_lm/conformer.py

```diff
@@ -0,0 +1,1484 @@
+# Copyright (c) 2021 University of Chinese Academy of Sciences (author: Han Zhu)
+# Apache 2.0
+
+import copy
+import math
+from typing import Dict, List, Optional, Tuple
+
+import torch
+import torch.nn as nn
+from torch import Tensor
+from torch.nn.utils.rnn import pad_sequence
+
+# Note: TorchScript requires Dict/List/etc. to be fully typed.
+Supervisions = Dict[str, torch.Tensor]
+
+
+class MaskedLmConformer(nn.Module):
+    def __init__(
+        self,
+        num_classes: int,
+        d_model: int = 256,
+        nhead: int = 4,
+        dim_feedforward: int = 2048,
+        num_encoder_layers: int = 6,
+        num_decoder_layers: int = 6,
+        dropout: float = 0.1,
+        cnn_module_kernel: int = 31,
+    ) -> None:
+        """
+        Args:
+          num_classes:
+            The input and output dimension of the model (inputs and outputs are
+            both discrete)
+          d_model:
+            Attention dimension.
+          nhead:
+            Number of heads in multi-head attention.
+            Must satisfy d_model // nhead == 0.
+          dim_feedforward:
+            The output dimension of the feedforward layers in encoder/decoder.
+          num_encoder_layers:
+            Number of encoder layers.
+          num_decoder_layers:
+            Number of decoder layers.
+          dropout:
+            Dropout in encoder/decoder.
+       """
+        super(MaskedLmConformer, self).__init__()
+
+        self.num_classes = num_classes
+
+        # self.embed is the embedding used for both the encoder and decoder.
+        self.embed_scale = d_model ** 0.5
+        self.embed = nn.Embedding(
+            num_embeddings=self.num_classes, embedding_dim=d_model,
+            _weight=torch.randn(self.num_classes, d_model) * (1 / self.embed_scale)
+        )
+
+        self.encoder_pos = RelPositionalEncoding(d_model, dropout)
+
+        encoder_layer = MaskedLmConformerEncoderLayer(
+            d_model,
+            nhead,
+            dim_feedforward,
+            dropout,
+            cnn_module_kernel,
+        )
+        self.encoder = MaskedLmConformerEncoder(encoder_layer, num_encoder_layers,
+                                                norm=nn.LayerNorm(d_model))
```

Comment:
We are using pre-normalization here.
You have placed a layer norm at the end of the encoder layer, see
https://github.com/k2-fsa/icefall/blob/0cfa8c80e0fad48e14d38784baa4802e0714f842/egs/librispeech/ASR/conformer_lm/conformer.py#L907

You are using an extra layer norm here, which means you are doing
```
x = layernorm(layernorm(x))
```

See
https://github.com/k2-fsa/icefall/blob/0cfa8c80e0fad48e14d38784baa4802e0714f842/egs/librispeech/ASR/conformer_lm/conformer.py#L967-L968

----

I just realized that it is even worse.

You are using a layer norm at both ends of an encoder layer but
encoder layers are connected end-by-end, which means the output of the layer norm
from the previous encoder layer is used as the input of the layer norm of the next encoder.

Reply:
This "even worse" part is not quite right, because there are bypass connections, so there are paths involving residuals where the input is used without layer norm.

Reply:
.. but yes, I agree that the LayerNorm at the end of the conformer encoder is redundant.  I have since stopped using that.  But in this particular case, it would take a long time to retrain the model if we were to fix it, so I'd say leave it as-is for now.